### PR TITLE
Support streaming responses.

### PR DESCRIPTION
**Pull request by alexmv**
Django streaming responses[1] return their "response," and then
repeatedly yield from their `.streaming_content` iterator to generate
the body.  Ending the request when the response is returned thus omits
all of the time spent streaming the response -- and by then the
HoneyDBWrapper's context has also been exited, which means any
database queries therein are also not traced.

Explicitly examine responses to determine if they are streaming --
and, if so, delay closing the root trace for the response until the
end of the streaming content.

[1] https://docs.djangoproject.com/en/3.2/ref/request-response/#streaminghttpresponse-objects

### CODE REVIEW

File: beeline/middleware/django/__init__.py [hunk 1]
```diff
@@ -158,6 +169,17 @@ def __call__(self, request):
                 for connection in connections.all():
                     stack.enter_context(connection.execute_wrapper(db_wrapper))
                 response = self.create_http_event(request)
+
+                def wrap_streaming_content(content):
```
Comment (ajvondrak) [inline beeline/middleware/django/__init__.py:173]: Could you elaborate on what the wrapping is achieving here? I think I can guess, but wanted to make sure I understood after our whole conversation in Slack. 🙂

File: beeline/middleware/django/test_django.py [hunk 2]
```diff
@@ -18,6 +18,7 @@ def test_call_middleware(self):
         ''' Just call the middleware and ensure that the code runs '''
         mock_req = Mock()
         mock_resp = Mock()
+        mock_resp.return_value.streaming = False
```
Comment (ajvondrak) [inline beeline/middleware/django/test_django.py:21]: Funny enough, when we were having our discussion in Slack, on of the first things I tried to do was crack open the tests and see if there was a way to exercise the case of a streaming response. I didn't get very far (I'm not that familiar with the mechanics of Django), but I think it's definitely the sort of thing we _should_ have tests around. Any ideas?

File: beeline/middleware/django/__init__.py (hunk 1 above)
Comment (alexmv) [reply beeline/middleware/django/__init__.py:173]: Yeah, this deserves a comment explaining it more -- the first `wrap_streaming_content`'s utility is more evident, but this one isn't.

The complexity is that `HoneyDBWrapper` only works for database queries run inside its `with contextlib.ExitStack()`.  Without this, the queries are run after the `response` is returned, well after the `contextlib.ExitStack()` has cleaned up the contexts that do the tracing.

So this `wrap_streaming_content` sets up a second context that runs later, in the context of when the generator is being called, that provides that same context again.  The first hunk above makes the request not end until the streaming is done; this one ensures that DB queries that happen while in that streaming section are recorded properly.

File: beeline/middleware/django/test_django.py (hunk 2 above)
Comment (alexmv) [reply beeline/middleware/django/test_django.py:21]: You are absolutely right with the nudge that the above code deserves a test. :)

I'll see if I have time to whip one up this evening.

**Review by vreynolds — APPROVED** [review]
Thank you!

**Comment by alexmv** [comment]
Yay!  Are you folks planning to do a release with this in it?  I'd love to deploy it without having to keep pinning the commit. :)

**Comment by paulosman** [comment]
> Yay! Are you folks planning to do a release with this in it? I'd love to deploy it without having to keep pinning the commit. :)

We do a sweep every other week or so - stay tuned!

**Pull request merged**